keyberon/src/layout.rs

```diff
@@ -1215,24 +1215,28 @@ impl<'a, const C: usize, const R: usize, const L: usize, T: 'a + Copy + std::fmt
             self.dequeue(queued);
         }
     }
-    fn press_as_action(&self, coord: KCoord, layer: usize) -> &'a Action<'a, T> {
+    fn resolve_coord(&self, coord: KCoord) -> &'a Action<'a, T> {
         use crate::action::Action::*;
-        let action = self
-            .layers
-            .get(layer)
-            .and_then(|l| l.get(coord.0 as usize))
-            .and_then(|l| l.get(coord.1 as usize));
-        match action {
-            None => &NoOp,
-            Some(Trans) => {
-                if layer != self.default_layer {
-                    self.press_as_action(coord, self.default_layer)
-                } else {
-                    &NoOp
+        for layer in self
+            .active_held_layers()
+            .iter()
+            .rev()
+            .chain(&[self.default_layer])
+        {
+            let action = self
+                .layers
+                .get(*layer)
+                .and_then(|l| l.get(coord.0 as usize))
+                .and_then(|l| l.get(coord.1 as usize));
+            match action {
+                None => return &NoOp,
+                Some(Trans) => {
+                    continue;
                 }
+                Some(action) => return action,
```

Comment:
This would still resolve to the same action if the original `Trans` is nested inside another action I think.

E.g. trans inside a tap-hold

Reply:
That's right. I haven't tried to resolve #635 in this PR, only #738.

Reply:
I haven't tested - away from desk - but wouldn't it loop while held and then eventually resolve the tap action when released? If correct, doing the tap seems undesirable.

Reply:
I went and tested it, and I was wrong about what the result would be. It turns out that the result is more fatal than an unintentional tap activation. I haven't dug fully into the reason, but seems that the state machine ends up in an infinite loop.

There probably needs to be some code that somehow knows about previously attempted layers to avoid this bad case. But the puzzle then is - how to know that the knowledge of previously attempted layers is up-to-date, and when is the right time to clear it? I don't have any answers for now.

Another alternative is to remove this change in `do_action` and keep the rest.

Reply:
I see. I've also tested with with `(multi _)` and it gives recursion overflow instead.

I've attempted to solve this in and decided that might as well implemented https://github.com/jtroo/kanata/issues/635 while doing that. See fe2e4f7. I'm saying *attempted* because it currently doesn't work, because of borrow checker doesn't like what I did. To me it seems like the right direction, but I thought I would ask for an second opinion before trying to fix the errors.

Reply:
Perhaps you haven't gotten around to it because of the borrow checking errors, but there currently appears to be no code that shrinks the number of available layers (e.g. in the `Trans` branch of `do_action`). Related, presently the `waiting_into_*` code looks problematic because fresh layer stacks are created, when instead the code should use an existing, smaller, layer stack.

As a note, the way I would avoid borrow checker errors is to undo the change I suggested earlier with `Iterator` - which was a suggestion made for simpler-looking code. One can still avoid heap allocations (to avoid more expensive allocator calls) by using an `ArrayDeque` that copies the layer indices. Possibly the `ArrayDeque` could be bounded by the `L` generic parameter.

---

Switching topics to the level of "good/bad idea", I wonder if instead of solving #635, we should deprecate (have warning logs for, and eventually error on) `Trans` actions nested within a more complicated action?

On the argument for deprecating, reading through ZMK / QMK docs, I don't see a way to use transparency behaviour within a more complicated action. It does seem to be a complicated use case. But I haven't played with QMK in a long time and never toyed with ZMK beyond a very basic level, so perhaps it is possible?

Arguing against deprecation, maybe it's another nice advantage and differentiation over these other projects, to support transparency in more places.

Reply:
> Perhaps you haven't gotten around to it because of the borrow checking errors, but there currently appears to be no code that shrinks the number of available layers (e.g. in the `Trans` branch of `do_action`).

No, it is supposed to be shrinked by iterating `layer_stack` in `resolve_coord`. Maybe it isn't obvious, because I forgot to add `&mut` to the passed iterator. Added `&mut` in 238cc9b.

> Related, presently the `waiting_into_*` code looks problematic because fresh layer stacks are created, when instead the code should use an existing, smaller, layer stack.

Oh, definitely that'll be the next thing to fix.

Reply:
> As a note, the way I would avoid borrow checker errors is to undo the change I suggested earlier with `Iterator` - which was a suggestion made for simpler-looking code. One can still avoid heap allocations (to avoid more expensive allocator calls) by using an `ArrayDeque` that copies the layer indices. Possibly the `ArrayDeque` could be bounded by the `L` generic parameter.

Fixed the errors in c510025. For now, changed to return `std::Vec`, for simplicity. The idea of shrinking layers with iterator haven't changed though, so instead of returning iterator right away, now a `Vec` is returned, which immediately is converted into iterator.

Reply:
alright, `_` is now implemented to work in actions. I've added quite a few test to check if it's working correctly.